### PR TITLE
[full-ci] Expiration date limits for reshare of a received share

### LINKED ISSUE #37013: Expiration date limits for reshare of a received share
Example: (today is 2020-02-25)

1) There is no default expiration date set (i.e. default is "never expire")
2) Anne shares `folder` with Bob and sets expiration date 2020-03-03
3) Bob shares with Carol and sets expiration date to 2020-03-10 (or clears the expiration date)
After 2020-03-03 Bob cannot see the folder, but Carol can see it. That seems odd, since Anne originally only gave Bob access until 2020-03-03, so why can Bob grant Carol longer access.

1) There is default expiration date set to e.g. 7 days (but not enforced)
2) Anne shares `folder` with Bob and expiration date defaults to 2020-03-03
3) Bob shares with Carol and sets expiration date to 2020-03-10 (or clears the expiration date)
After 2020-03-03 Bob cannot see the folder, but Carol can see it. That seems odd, since Anne originally only gave Bob access until 2020-03-03, so why can Bob grant Carol longer access.

1) There is default expiration date set to e.g. 7 days and enforced
2) Anne shares `folder` with Bob and sets expiration date to 2020-02-28 (earlier than the enforced maximum)
3) Bob shares with Carol and sets expiration date to 2020-03-03 (the enforced maximum)
After 2020-02-28 Bob cannot see the folder, but Carol can see it. That seems odd, since Anne originally only gave Bob access until 2020-02-28, so why can Bob grant Carol longer access.

IMO when I receive a share that has an expiration date set, and has reshare permission, then I should have to keep and expiration date when I reshare it, and the expiration date of the reshare should be <= the expiration date of the received share. (i.e. I should not be able to grant longer access than the access that I received)

@pmaier1 or... what is the required behaviour?

### PR DESCRIPTION
**Pull request by AlexAndBear**
## Description
With this change, if a user re-shares an existent share with an expiration date,
the expiration date of the target share will be taken for the new share,
if the user doesn't provide an expiration date.

If the expiration date is provided and exceeds the date of the target share,
an error will be thrown.

## Related Issue

- Fixes https://github.com/owncloud/core/issues/37013

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiShareReshareToRoot2-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/33554/75/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiShareUpdateToShares-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/33556/80/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiShareReshareToRoot2-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/33567/75/1

File: tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature [hunk 1]
```diff
@@ -370,7 +370,7 @@ Feature: resharing a resource with an expiration date
       | 1               | no                  | no                  | 100             |
       | 2               | no                  | no                  | 200             |
 
-  @skipOnOcV10 @skipOnOcV10.3 @issue-37013
+  @skipOnOcV10.6 @skipOnOcV10.7 @skipOnOcV10.8
   Scenario Outline: reshare extends the received expiry date up to the default by default
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature:374]:
```suggestion
  Scenario Outline: reshare keeps the received expiry date by default
```

File: tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature [hunk 2]
```diff
@@ -407,7 +407,7 @@ Feature: resharing a resource with an expiration date
       | 1               | no                  | no                  | 100             |
       | 2               | no                  | no                  | 200             |
 
-  @skipOnOcV10 @skipOnOcV10.3 @issue-37013
+  @skipOnOcV10.6 @skipOnOcV10.7 @skipOnOcV10.8
   Scenario Outline: reshare can extend the received expiry date further into the future
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature:411]:
```suggestion
  Scenario Outline: reshare cannot extend the received expiry date further into the future
```

**Review by phil-davis — COMMENTED** [review]
Just a few words of scenario description that can be changed next time you are touching the code.

**Comment by mmattel** [comment]
@JanAckermann any progress on that? Seems a reasonable change that could go into 10.11

File: tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature [hunk 3]
```diff
@@ -403,19 +403,19 @@ Feature: resharing a resource with an expiration date
...
       | expiration | +20 days |
     Examples:
-      | ocs_api_version | default-expire-date |
-      | 1               | yes                 |
-      | 2               | yes                 |
-      | 1               | no                  |
-      | 2               | no                  |
+      | ocs_api_version | default-expire-date | ocs_status_code | http_status_code |
+      | 1               | yes                 | 996             | 200              |
+      | 2               | yes                 | 500             | 500              |
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature:414]: ToDo: we should not expect a 500 HTTP or OCS status code. Needs to be sorted out.

File: tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature [hunk 4]
```diff
@@ -403,19 +403,19 @@ Feature: resharing a resource with an expiration date
...
+      | 1               | no                  | 996             | 200              |
+      | 2               | no                  | 500             | 500              |
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature:416]: ToDo: we should not expect a 500 HTTP or OCS status code. Needs to be sorted out.

File: tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature [hunk 5]
```diff
@@ -435,11 +435,11 @@ Feature: resharing a resource with an expiration date
...
+    And the OCS status code should be "<ocs_status_code>"
     And the information of the last share of user "Alice" should include
       | expiration | +20 days |
     Examples:
-      | ocs_api_version |
-      | 1               |
-      | 2               |
+      | ocs_api_version | default-expire-date | ocs_status_code | http_status_code |
+      | 1               | yes                 | 996             | 200              |
+      | 2               | yes                 | 500             | 500              |
```
Comment (phil-davis) [inline tests/acceptance/features/apiShareReshareToShares3/reShareWithExpiryDate.feature:445]: ToDo: we should not expect a 500 HTTP or OCS status code. Needs to be sorted out.

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiShareReshareToRoot2-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/36072/76

**Comment by phil-davis** [comment]
I rebased this in #40263 - if that works nicely, we can probably close this old PR.

**Pull request closed without merging**